### PR TITLE
feat(elementhandle): introduce elementHandle.isIntersectingViewport() method.

### PR DESCRIPTION
**Pull request by b-ponomarenko**
Add API to define is element visible in viewport via Intersection Observer

References #2629

### CODE REVIEW

**Comment by b-ponomarenko** [comment]
I signed it!

**Comment by b-ponomarenko** [comment]
Can I rerun `node6 (windows)`?

**Comment by b-ponomarenko** [comment]
@aslushnikov please check

File: docs/api.md [hunk 1]
```diff
@@ -219,6 +219,7 @@
   * [elementHandle.getProperties()](#elementhandlegetproperties)
   * [elementHandle.getProperty(propertyName)](#elementhandlegetpropertypropertyname)
   * [elementHandle.hover()](#elementhandlehover)
+  * [elementHandle.isDisplayed()](#elementhandleisdisplayed)
```
Comment (JoelEinbinder) [inline docs/api.md]: nit: I like isVisible over isDisplayed, because we already use the word visible in waitForSelector.

Comment (aslushnikov) [inline docs/api.md]: `elementHandle.isVisible()` is better since it aligns with `waitForSelector`'s "visible" option.

File: lib/ElementHandle.js [hunk 2]
```diff
@@ -341,6 +341,24 @@ class ElementHandle extends JSHandle {
...
+   * @returns {!Promise<boolean>}
+   */
+  isDisplayed() {
+    return this.executionContext().evaluate(
+        node => new Promise(resolve => {
+          const callback = entries => {
+            resolve(entries[0].isIntersecting);
+            observer.unobserve(node);
+          };
+          const observer = new IntersectionObserver(callback, { root: null, rootMargin: '0px', threshold: 1.0 });
```
Comment (aslushnikov) [inline lib/ElementHandle.js]: so when will the `callback` be called? Is it always called initially?

File: test/elementhandle.spec.js [hunk 3]
```diff
@@ -184,6 +184,17 @@ module.exports.addTests = function({testRunner, expect}) {
...
 
+  describe('ElementHandle.isDisplayed', function() {
+    it('should return false if element is not visible in viewport', async({page, server}) => {
+      await page.goto(server.PREFIX + '/hidden-and-visible.html');
+      await page.setViewport({ width: 1000, height: 400 });
+
+      const box = await page.$('.box');
+
+      expect(await box.isDisplayed()).toBe(false);
+    });
```
Comment (aslushnikov) [inline test/elementhandle.spec.js]: can we also add a test that checks for the truthy value?

File: test/assets/hidden-and-visible.html [hunk 4]
```diff
@@ -0,0 +1,23 @@
+<html>
```
Comment (aslushnikov) [inline test/assets/hidden-and-visible.html]: let's use `page.setContent` instead of a separate asset.

**Review by aslushnikov — COMMENTED** [review]
Thanks, this looks promising

As part of this PR, let's change `waitForSelector` implementation to use intersection observer; this way we won't have inconsistencies like this:

```js
const e = await page.waitForSelector('div', {visible: true});
await e.isVisible(); // false
```

**Comment by JoelEinbinder** [comment]
If we change waitForSelector to use intersection observer, won’t that cause elements below the fold to be considered invisible? That’s a pretty major change, and I’m not sure if it makes that feature more or less useful. “Visible” vs “would visible if you scroll down” is hard to define. And it conflicts with the original use case for this feature, which was to report elements ina carousel as invisible even though they could be scrolled to horizontally.

**Comment by aslushnikov** [comment]
> won’t that cause elements below the fold to be considered invisible?

@JoelEinbinder These will be considered invisible. This is a very good point though, we might need to postpone this till 2.0. 

> That’s a pretty major change, and I’m not sure if it makes that feature more or less useful.

Usefulness is debatable, but API sanity will improve:
- Our current `isVisible` is more like `isClickable`, it's quite quirky with a huge grey area of what is considered "visible".
- Visibility defined by intersection observer is aligned with a natural perception of visibility, it's easy to explain and understand.

File: docs/api.md (hunk 1 above)
Comment (b-ponomarenko) [reply docs/api.md]: Ok, I will rename to `isVisible`

File: lib/ElementHandle.js (hunk 2 above)
Comment (b-ponomarenko) [reply lib/ElementHandle.js]: Yes, it will be called initially and then unobserve from the node

File: test/elementhandle.spec.js (hunk 3 above)
Comment (b-ponomarenko) [reply test/elementhandle.spec.js]: Yes, I'll add this test case soon

**Comment by b-ponomarenko** [comment]
@aslushnikov change implementation of `_waitForSelectorOrXPath`, but in some cases [this test](https://github.com/GoogleChrome/puppeteer/blob/master/test/frame.spec.js#L287) fails (when I run this test in single mode almost always test passed), what can be wrong?

**upd**
no longer relevant, understood the problem and fix

**Comment by b-ponomarenko** [comment]
@aslushnikov please check PR

**Comment by aslushnikov** [comment]
> @aslushnikov please check PR

@b-ponomarenko We discussed this PR with @JoelEinbinder one more time.

Summary:
- there are two types of visibility: visible-in-viewport (this implementation) and displayed-in-page (currently used for `visible` option in `waitForSelector`)
- both make perfect sense from API standpoint: visible-in-viewport comes handy to detect what user sees, and displayed-in-page is more like "interactable", e.g. something that could be clicked.
- changing `waitForSelector`'s `visible` semantics from displayed-in-page to visible-in-viewport is a **breaking change**; we can't afford it yet.

So the way forward with this:
- disregard my suggestion on using IntersectionObserver for `waitForSelector`. (Sorry for misguiding you)
- introduce a new distinct method on `ElementHandle`, e.g. `ElementHandle.isVisibleInViewport`, that uses Intersection Observer.

As a part of [2.0 effort](https://github.com/GoogleChrome/puppeteer/issues/2079), we might want to do a better job separating "interactivity" vs "visibility", e.g. renaming `visible` into `interactable`. I've added a bullet.

**Comment by b-ponomarenko** [comment]
@aslushnikov I canceled using Intersection Observer in `_waitForSelectorOrXPath`. What's next? Really, maybe rename `visible` property to `interactable` or some other property name, and move current implementation via Intersection Observer to `visible` property?

**Review by aslushnikov — APPROVED** [review]
We discussed this with @JoelEinbinder and renamed into `elementHandle.isIntersectingViewport()`. Thanks for the PR!

**Comment by justnpT** [comment]
I wouldn't be so sure that inputElement.isIntersectingViewport() is true only when an element is visible for human eyes, I have a case where element didn't yet appear as visible for eyes and
`await element.isIntersectingViewport()`
returns true

**Comment by aslushnikov** [comment]
>  I have a case where element didn't yet appear as visible for eyes and
await element.isIntersectingViewport()
returns true

@justnpT interesting! Can you please file a new issue with a repro for your usecase?

**Pull request merged**